petstore/src/main/scala/io/finch/petstore/endpoint.scala

```diff
@@ -0,0 +1,206 @@
+package io.finch.petstore
+
+import com.twitter.finagle.Service
+import com.twitter.finagle.httpx.{Request, Response}
+import com.twitter.util.{Future}
+import io.finch.argonaut._
+import io.finch.request._
+import io.finch.response._
+import io.finch.route._
+
+/**
+ * Provides the paths and endpoints for all the API's public service methods.
+ */
+object endpoint extends ErrorHandling {
+
+  /**
+   * Private method that compiles all pet service endpoints.
+   * @return Bundled compilation of all pet service endpoints.
+   */
+  private def pets(db: PetstoreDb) =
+    getPet(db) :+:
+    addPet(db) :+:
+    updatePet(db) :+:
+    getPetsByStatus(db) :+:
+    findPetsByTag(db) :+:
+    deletePet(db) :+:
+    updatePetViaForm(db) :+:
+    uploadImage(db)
+
+  /**
+   * Private method that compiles all store service endpoints.
+   * @return Bundled compilation of all store service endpoints.
+   */
+  private def store(db: PetstoreDb) =
+    getInventory(db) :+:
+    addOrder(db) :+:
+    deleteOrder(db) :+:
+    findOrder(db)
+
+  /**
+   * Private method that compiles all user service endpoints.
+   * @return Bundled compilation of all user service endpoints.
+   */
+  private def users(db: PetstoreDb) =
+    addUser(db) :+:
+    addUsersViaList(db) :+:
+    addUsersViaArray(db) :+:
+    getUser(db) :+:
+    deleteUser(db) :+:
+    updateUser(db)
+
+  /**
+   * Compiles together all the endpoints relating to public service methods.
+   * @return A service that contains all provided endpoints of the API.
+   */
+  def makeService(db: PetstoreDb): Service[Request, Response] = handleExceptions andThen (
+    pets(db) :+:
+    store(db) :+:
+    users(db)
+  ).toService
+
+  /**
+   * The long passed in the path becomes the ID of the Pet fetched.
+   * @return A Router that contains the Pet fetched.
+   */
+  def getPet(db: PetstoreDb): Router[Future[Pet]] = Get / "pet" / long /> db.getPet
+
+  /**
+   * The pet to be added must be passed in the body.
+   * @return A Router that contains a RequestReader of the ID of the Pet added.
+   */
+  def addPet(db: PetstoreDb): Router[RequestReader[Long]] = Post / "pet" /> body.as[Pet].embedFlatMap(db.addPet)
+
+  /**
+   * The updated, better version of the current pet must be passed in the body.
+   * @return A Router that contains a RequestReader of the updated Pet.
+   */
+  def updatePet(db: PetstoreDb): Router[RequestReader[Pet]] = Put / "pet" /> {
+    body.as[Pet].embedFlatMap { pet =>
+      for {
+        _ <- db.updatePet(pet.copy(id = Some(pet.id.getOrElse(-1))))
```

Comment:
It's a good principle to let the type system help you whenever you can (and something we especially want to do here since this is an example project). The type system doesn't prevent a pet with an `id` of `-1`, in which case this reasoning would break down.  Handling a `None` value is also just a little clearer.